File: stanford/traveling_salesman_heuristic/julia/heap.jl
function insert(edge::Int32, array::Array{Int32, 1})
  push!(array, edge)
  heapify_up(array)
end

function extract_min(array::Array{Int32, 1})
  array[1], array[length(array)] = array[length(array)], array[1]
  min_edge = array[length(array)]
  heapify_down(array)
  return min_edge
end

function heapify_up(array::Array{Int32, 1})
  parent_idx = find_parent_idx(new_edge_idx)
  while new_edge_idx != 1 && array[new_edge_idx] < array[parent_idx,3]
    array[new_edge_idx], array[parent_idx] = array[parent_idx], array[new_edge_idx]
    new_edge_idx = parent_idx
    parent_idx = find_parent_idx(new_edge_idx)
  end
end

function heapify_down(array::Array{Float64, 2})
  current_edge_idx = 1
  swap = true
  while swap
    swap = false
    children = find_children_indeces(current_edge_idx)
    swap_idx = current_edge_idx
    for child_idx in children
      if array[child_idx] <= array[swap_idx]
        swap_idx = child_idx
        swap = true
      end
    end
    if swap
      array[current_edge_idx], array[swap_idx] = array[swap_idx], array[current_edge_idx]
      current_edge_idx = swap_idx
    end
  end
end

function find_parent_idx(idx::Int64)
  return floor(Int,(idx - 1) / 2)
end

function find_children_indeces(idx::Int64, array_length::Int64)
  result = Array{Int64,1}(0)
  left = (2 * idx) + 1
  right = (2 * idx) + 2
  if left < array_length
    push!(result, left)
    if right < array_length
      push!(result, right)
    end
  end
  return result
end

function calculate_distance(coord1::Array{Float64,1}, coord2::Array{Float64,1})
  return sqrt((coord1[1] - coord2[1])^2 + (coord1[2] - coord2[2])^2)
end

File: stanford/traveling_salesman_heuristic/py/heap.py
import numpy as np

class MinHeap(object):
    def __init__(self):
        self.store = []

    def insert(self, edge):
        self.store.append(edge)
        self.heapify_up()

    def extract_min(self):
        self.store[0], self.store[-1] = self.store[-1], self.store[0]
        min_edge = self.store.pop()
        self.heapify_down()
        return min_edge

    def heapify_up(self):
        new_edge_idx = len(self.store) - 1
        parent_idx = int((new_edge_idx - 1) / 2)
        while new_edge_idx != 0 and \
        self.store[new_edge_idx][2] <= self.store[parent_idx][2]:
            if self.store[new_edge_idx][2] == self.store[parent_idx][2]:
                if self.store[new_edge_idx][1] > self.store[parent_idx][1]:
                    break
            self.store[new_edge_idx], self.store[parent_idx] = \
            self.store[parent_idx], self.store[new_edge_idx]
            new_edge_idx = parent_idx
            parent_idx = int((new_edge_idx - 1) / 2)

    def heapify_down(self):
        current_edge_idx = 0
        swap = True
        while swap:
            swap = False
            children = self.children_indeces(current_edge_idx)
            swap_idx = current_edge_idx
            for child_idx in children:
                if self.store[child_idx][2] <= self.store[swap_idx][2]:
                    if self.store[child_idx][2] == self.store[swap_idx][2]:
                        if self.store[child_idx][1] > self.store[swap_idx][1]:
                            continue
                    swap_idx = child_idx
                    swap = True
            if swap:
                self.store[current_edge_idx], self.store[swap_idx] = \
                self.store[swap_idx], self.store[current_edge_idx]
                current_edge_idx = swap_idx
        return self.store

    def children_indeces(self, parent_idx):
        result = []
        length = len(self.store)
        left = (2 * parent_idx) + 1
        right = (2 * parent_idx) + 2
        if left < length:
            result.append(left)
            if right < length:
                result.append(right)
        return result
